### PR TITLE
Improve usability of ViewLocalizer

### PR DESCRIPTION
**Pull request by DamianEdwards**
Changes to `ViewLocalizer` resource lookup semantics:
- Now uses `ViewContext.ExecutingFilePath` before `ViewContext.View.Path` for determining base name
- Trims any file extension from the base name
- Always prepends the application name (note the `ResourceManagerStringLocalizerFactory`'s new behavior WRT the application name prefix makes this all work well)
- Updated the loc sample app to match (also used by the functional tests)
- Added unit tests to cover this behavior
- #3718
- #2767

@ryanbrandenburg @rynowak

### CODE REVIEW

File: src/Microsoft.AspNet.Mvc.Localization/IViewLocalizer.cs [hunk 1]
```diff
@@ -4,7 +4,7 @@
 namespace Microsoft.AspNet.Mvc.Localization
 {
     /// <summary>
-    /// A service that provides localized strings for views.
+    /// Represents a type that provides HTML-aware localization for views.
```
Comment (ryanbrandenburg) [inline src/Microsoft.AspNet.Mvc.Localization/IViewLocalizer.cs]: Is this our usual wording? Seems kinda off.

File: src/Microsoft.AspNet.Mvc.Localization/ViewLocalizer.cs [hunk 2]
```diff
@@ -93,12 +96,30 @@ public void Contextualize(ViewContext viewContext)
...
+
+            // Trim the file extension from the end of the path
+            if (!string.IsNullOrEmpty(path) && Path.HasExtension(path))
             {
-                baseName = baseName.Substring(1);
+                var extension = Path.GetExtension(path);
+                path = path.Substring(0, path.Length - extension.Length);
             }
 
+            Debug.Assert(!string.IsNullOrEmpty(path), "Couldn't determine a path for the view");
```
Comment (ryanbrandenburg) [inline src/Microsoft.AspNet.Mvc.Localization/ViewLocalizer.cs]: Just a reminder that this is here, fine to leave it but don't know if it's intentional.

Comment (DamianEdwards) [inline src/Microsoft.AspNet.Mvc.Localization/ViewLocalizer.cs]: It was intentional.

File: src/Microsoft.AspNet.Mvc.Localization/IViewLocalizer.cs (hunk 1 above)
Comment (DamianEdwards) [inline src/Microsoft.AspNet.Mvc.Localization/IViewLocalizer.cs]: This is standard .NET API reference docs language, e.g. https://msdn.microsoft.com/en-us/library/s4ys34ea(v=vs.110).aspx

File: src/Microsoft.AspNet.Mvc.Localization/ViewLocalizer.cs [hunk 3]
```diff
@@ -93,12 +96,30 @@ public void Contextualize(ViewContext viewContext)
...
 
-            var baseName = viewContext.View.Path.Replace('/', '.').Replace('\\', '.');
-            if (baseName.StartsWith(".", StringComparison.OrdinalIgnoreCase))
+            // Given a view path "/Views/Home/Index.cshtml" we want a baseName like "MyApplication.Views.Home.Index"
+
+            var path = viewContext.ExecutingFilePath;
+
+            if (string.IsNullOrEmpty(path))
+            {
+                path = viewContext.View.Path;
```
Comment (ryanbrandenburg) [inline src/Microsoft.AspNet.Mvc.Localization/ViewLocalizer.cs]: Love this change, having the "cshtml" in there always struck me as super weird.

**Comment by ryanbrandenburg** [comment]
:shipit: from me since you addressed all my concerns.

File: src/Microsoft.AspNet.Mvc.Localization/ViewLocalizer.cs [hunk 4]
```diff
@@ -93,12 +96,30 @@ public void Contextualize(ViewContext viewContext)
...
+            }
+
+            // Trim the file extension from the end of the path
+            if (!string.IsNullOrEmpty(path) && Path.HasExtension(path))
```
Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.Localization/ViewLocalizer.cs]: Might want to consider using a `StringBuilder` to do these replaces \ substrings if it happens in ordinary cases.

Comment (DamianEdwards) [inline src/Microsoft.AspNet.Mvc.Localization/ViewLocalizer.cs]: Too late. Log an issue?

Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.Localization/ViewLocalizer.cs]: Sure. Depends on how common these cases are. Although my guess is nearly all view paths start with a leading slash and have an extension so it's not uncommon.

**Pull request merged**